compiler/rustc_ast/src/token.rs

```diff
@@ -529,11 +529,15 @@ impl Token {
     }
 
     /// Returns `true` if the token can appear at the start of a generic bound.
+    // FIXME(negative_bounds): Ideally this would contain `!` (`Not`), too,
+    // but that would regress pre-2018 `dyn!(…)` (macro calls).
     pub fn can_begin_bound(&self) -> bool {
         self.is_path_start()
             || self.is_lifetime()
-            || self.is_keyword(kw::For)
             || self == &Question
+            || self == &Tilde
+            || self.is_keyword(kw::Const)
+            || self.is_keyword(kw::For)
```

Comment:
Why do we need to change this? This is just going to cause things like bare trait object `t: const Foo` work, right?

Reply:
No, the parser uses `Parser::can_begin_bound` for bare trait object types, so `let _: const Trait;` would continue to parse even if I removed `kw::Const` from `Token::can_begin_bound`.
`Token::can_begin_bound` is used for only used for `dyn`-prefixed trait object types and allows us to parse `dyn const Trait` otherwise one'd need to write `dyn (const Trait)` (ignoring the fact that we reject dyn-const later during AST validation).

Reply:
Why are there two `can_begin_bound`s :(

Reply:
Wait --

```
/// Is a `dyn B0 + ... + Bn` type allowed here?
    fn is_explicit_dyn_type(&mut self) -> bool {
        self.check_keyword(kw::Dyn)
            && (self.token.uninterpolated_span().at_least_rust_2018()
                || self.look_ahead(1, |t| {
                    (t.can_begin_bound() || t.kind == TokenKind::BinOp(token::Star))
                        && !can_continue_type_after_non_fn_ident(t)
                }))
    }
```

Wait, leaving this out only breaks `dyn const Trait` in edition 2015, though, right?

Reply:
However, updating *does* mean that the following stable code regresses:

```rs
macro_rules! check {
    ($ty:ty) => { compile_error!(""); };
    (dyn $c:ident) => {  };
}

check! { dyn const }

fn main() {}
```

which is an alteration of [the regression found by fee1-dead](https://github.com/rust-lang/rust/pull/119099#discussion_r1436021036).

Reply:
I'm suggesting to leave `Token::can_begin_bound()` alone, since it seems to only affect "`dyn const Trait`" parsing in edition 2015, since otherwise `Span::at_least_rust_2018()` will be true in `Parser::is_explicit_dyn_type` and we short-circuit.

Reply:
Alright, I'm gonna implement your suggestion since it does theoretically prevent more regressions, although I'm not super convinced that it's worth the loss of uniform behavior across editions.

Reply:
I'd argue that it's incredibly normal for new syntax addition to be edition-dependent due to grammar backcompat.

Reply:
Also, can you rename can_begin_bound to something like can_begin_dyn_bound_in_edition_2015 and remove any comments about keeping it in sync, since that ship has sailed?